### DISCUSSION TITLE
Modelling the synchromization between machines

### DISCUSSION THREAD

**Discussion by VanTanev** [statelyai/xstate#1708] [Q&A]
Hello, 

I'm hoping for some advice on how to approach a modelling problem I have.

I have a page that displays several records, that can be in either `readonly` or `edit` mode. I want to coordinate that mode switch from a central machine.

The record machines look something like this: 
```javascript
// https://xstate.js.org/viz/?gist=623e8a02dfeb5e4617c403fc93854076
const recordMachine = createMachine({
    id: 'record',
    initial: 'reading',
    context: {
      data: 'initialData'
    },
    states: {
      reading: {
        on: {
          ATTEMPT_EDIT: 'aquiringLock'
        }
      },
      aquiringLock: {
        invoke: {
          id: 'aquire-lock',
          src: 'postAqireLock',
          onDone: 'editing',
          onError: 'reading'
        }
      },
      editing: {
        initial: 'idle',
        states: {
          idle: {
            on: {
              UPDATE: 'updating',
              RELEASE: 'releasing'
            }
          },
          updating: {
            invoke: {
              id: 'update-data',
              src: 'putUpdateData',
              onDone: 'idle',
            }
          },
          releasing: {
            invoke: {
              id: 'release-lock',
              src: 'deleteLock',
              onDone: 'released'
            }
          },
          released: {
            type: 'final'
          },
        },
        onDone: 'reading'
      },
    }
  })
```

I need to write some sort of a coordinator machine, that can have an arbitrary number of record machines spawned inside it.
The user needs to be able to select a random selection of those, and change their state simultaneously.

That is to say, say there are three records - the user can choose 2 and say "attempt to edit these two." The result of this action must be either a success of all records locked for edit, or if any one fails, the successful ones need to be unlocked and the whole operation should fail.

This sounds like I need to spawn a machine as an invoke in my coordinator machine, give it the refs to the desired record machines, and it needs to do something like:
```
 -> send ATTEMPT_EDIT to all machines
 -> keep track of which have responded in internal context
 -> when all have responded (how? will sendParent even work to send to two different parents?)
 (if all successful) -> complete with a success, 
      and allow the coordinator to transition to a global `editing` state
 (if any failed) -> go to a cleanup phase where it sends `RELEASE` commands to any machines that succeeded, 
      and then complete with a failure.
```

Any guidance would be appreciated!

**Comment by davidkpiano** [comment]
My initial thoughts:

> -> send ATTEMPT_EDIT to all machines

Use the `actions.pure()` action creator:

```js
actions: actions.pure(context => {
  return context.records.map(record => {
    return actions.send('ATTEMPT_EDIT', { to: record });
  });
})
```

> -> keep track of which have responded in internal context

You'd have some `context.respondedRecords = {}` which can be something that tracks the responded state of each `record` actor. When a `record` responds (via `sendParent`), that event will cause that `respondedRecords` state to be updated.

> -> when all have responded (how? will sendParent even work to send to two different parents?)

This is a good use-case for an eventless transition. Think of it like "always check this".

```js
always: [/* see below */]
```

> (if all successful) -> complete with a success, 
>      and allow the coordinator to transition to a global `editing` state
> (if any failed) -> go to a cleanup phase where it sends `RELEASE` commands to any machines that succeeded, 
>      and then complete with a failure.

```js
always: [
  { target: 'editing', cond: 'allSuccessful' },
  { target: 'cleanup', cond: 'anyFailed', actions: 'releaseRecords' }
]
```

Where the `'releaseRecords'` action is also a `pure()` action:

```js
// machine options
actions: {
  releaseRecords: actions.pure(context => {
    // just like before
    return context.records.map(record => {
      return records.send('RELEASE', { to: record });
    }
  })
}
```

Hope that's enough to get you started! With state machines, there's usually many ways to model the logic, but this is what make the most sense to me. You might end up with something different, and that's fine.

**Reply by VanTanev** [reply]
Thanks for the pointers! Now I have most of the happy path down. 

Is there a way for an invoked machine to communicate back some data upon its completion? When `invoke` is used with a machine `src`, it seems that `onDone` gets an event with empty `data`.

To be more specific, what I'm trying to do is:

```
 - coordinator goes into "attemptEditing" state and spawns the "multiLockMachine"
 - multiLock machine has 2 final states: "success" and "failure", if it managed to lock all resources or not
 - I would like to communicate the final state back to the coordinating machine. 
 - based on final state, `onDone` uses action guards to select either `editing` state or go back to  `idle`  state
```

Overall, I think that if the `done.invoke.multi-lock-aquire` event given to `onDone()` had either the final `state` of the invoked machine, or the final `context`, that would be great.

**Reply by davidkpiano** [reply]
A final state can have `data` which is sent back in the `doneInvoke` event:

```js
success: {
  type: 'final',
  data: ctx => ctx // just send the entire context back
}
```

**Reply by VanTanev** [reply]
Oh, that's great! I was looking for something like this in the docs, but it seems to be missing?

**Reply by davidkpiano** [reply]
It's here: https://xstate.js.org/docs/guides/communication.html#done-data

**Reply by VanTanev** [reply]
Something that's unclear to me, what are the synchronization limitations between machines? What does `{ sync: true }` on a `spawn()` do, exactly?

For example, if I am a parent and I have references to child machines, is it safe to use those references in, eg, event guards?
```typescript
{
  initial: 'idle',
  entry: assign((ctx) => ({ 
    subMachineRef: spawn(someSubMachine)
  })),
  on: {
    TRANSITION: {
      cond: (ctx) => ctx.subMachineRef.state.matches('some state'), // is this unsafe without `{ sync: true }`?
      target: 'guarded'
    }
  },
  states: {
    idle: {},
    guarded: {},
  }
}
```
I would imagine that if `subMachineRef` is a true ref to some object, the above should work without `sync: true`.

**Reply by VanTanev** [reply]
@davidkpiano I'm very curious if the above is broken or not. I /think/ it should work, if I'm reading the code correctly (`subMachineRef` should be a service that is be the result of `interpret()`, and those should always have access to the current internal state), but the documentation seems to contradict me, so I'm confused.

**Reply by davidkpiano** [reply]
It might technically work because you're essentially reading mutable state, but that's the problem; you're reading mutable state.

XState v5 will make this state private, to more closely adhere to the actor model. It's better for that `subMachineRef` to directly send updates to the parent. Think of it like you're interacting with an API: you're not always going to get every single state change from the API to your frontend, nor are you going to be able to read the state directly (you have to send a GET request and wait for a response).

So for now, use `{ sync: true }` because that's essentially the same as the child machine sending the parent a notification that its state has updated. Here's an example of where reading mutable state will fail:

```js
always: {
  target: 'somewhere',
  cond: ctx => ctx.subMachineRef.state.matches('whatever')
}
```

You would expect the above transition to be taken when the `subMachineRef` state changes, _but it won't_ because the parent simply doesn't know that `subMachineRef`'s state changed.

**Reply by VanTanev** [reply]
Ah, this clears it up somewhat! So, speaking of V4:

 - An externally triggered event transition on the parent machine will be able to read the mutable state of a child machine. That is to say, if I `send()` an event to the parent machine, it can read the current mutable state of a child machine.
 - The parent machine's state however is not re-evaluated automatically when a child's state changes.
 - Because of this, `always` eventless transitions simply do not execute. /If/ they executed, they would be able to see the current mutable state of a child, but there is nothing to make the parent machine execute, unless we force that.
 - We can force the parent machine to execute, either by using `spawn()` with `{ sync: true }`, which just sends a `xstate.update` event to the parent machine on each child state transition, or with manual action calls to `sendParent()` in the child machine.

I'm not sure I understand how that would work in V5, then. Would the parent machine never have access to a spawned child's state, and we would only have true actor-like communication with events? How would we do a guard based on the state of a child machine then, something like:

```typescript
const parentMachine = createMachine({
  entry: assign(() => ({ 
    childRef: spawn(childMachine),
  })),
  initial: 'idle'
  states: {
    idle: {
       on: { CHECK: 'checking_child' }
    },
    checking_child: {
       entry: send('QUERY', { to: (ctx) => ctx.childRef })
       on: {
          RESPONSE: {
            actions: [{ cond: (_, e) => e.data, target: 'child_success' }, { target: 'child_fail' }]
          }
       }
    },
    child_success: {},
    child_fail: {},
  },
})

const childMachine = createMachine({
  context: {
    data: Math.random() > 0.5   
  },
  on: {
    QUERY: {
      actions: (ctx) => respond({ type: 'RESPONSE', data: ctx.data })
    }
  }
})
```

**Reply by davidkpiano** [reply]
> How would we do a guard based on the state of a child machine then, ...

No such thing as "based on the state"; everything is based on events. If you want to guard based on a child's state, that child needs to send the relevant part of its state back to the parent.

If you think about it, this is how we write programs that need to communicate with other APIs anyway. For example, this doesn't make sense:

```js
fetch('some/api').then(res => res.json())
  .then(response => {
    // ❌ can't magically read the server state directly!
    if (getStateFromServerSomehow().status === 'valid') {
      // ...
    }

    // ✅ it needs to be sent back in the response instead:
    if (response.status === 'valid') {
      // ...
    }
  });
```

**Reply by VanTanev** [reply]
@davidkpiano I understand what you mean, but it feels like for some cases it's introducing extra friction. I think we have some incompatible design considerations at hand, and there will be tradeoffs based on the API that xstate decides on.

 - OOP-like: If a parent machine spawns children and has references to them, it makes some sense to be able to use those references directly. This is literally ruby OOP - send message to an object, and the object responds with some value (or an exception).
 - Actor-like: The difference from above is mostly that we don't use the JS object ref to directly query its internal state, but we send a message and instead of receiving some value, we setup a listener for a response message.

In cases where we have a reference to the child machine already, to me it feels a little weird to have to go down the actor route. If XState decides to force that route, there should be some helpers for common use cases like this.

I think one of the most common cases will be, a parent must guard against a state transition (or, alternatively, /must/ execute a state transition) based on a child's state. In my use case, this manifests in two ways:

1. Parent wants to move all its children from state X to state Y, but should do abort if children are in state Z. If aborted, children that started in state X must at the end be back to state X. Possible approaches:

OOP-like :
```typescript
{
  states: {
    idle: {
        on: {
          TRANSITION_CHILDREN:  { 
            // if all children are in state X
            cond: (ctx) => ctx.childMachineRefs.every(child => child.state.matches('X')),
            // tell them to transition to state Y
            actions: pure((ctx) => ctx.childMachineRefs.map(child, send('GO_TO_Y', { to: child } }))
        }
   }
}
```

Actor-like:
```typescript
// parent
{
  states: {
    idle: {
        on: {
          TRANSITION_CHILDREN:  { 
            actions: [
              // create a list of pending responses
              assign(ctx => ({ 
                transitionResponses: ctx.chidMachineRefs.map((ref) => ({ id: ref.id, status: 'pending' })),
              })),
              // tell all children to attempt to transition to Y, and expect a 'RESPONSE' event with the result 
              pure((ctx) => ctx.childMachineRefs.map(child, send('GO_TO_Y', { to: child } })),
            ],
            target: 'waitResponse'
        }
   },
   waitResponse: {
      on: {
         RESPONSE: {
            actions: assing({
              // for each response, update our local response state status
              transitionResponses: (ctx, e) => ctx.transitionResponses.map(resp => resp.id === e.id ? {...resp, status: e.status} : resp)
         }
      },
      always: [{
          // if everyone is successful, then we completed the operation sucessfully
          cond: ctx => ctx.transitionResponses.every(resp => resp.status === 'success'),
          target: 'idle'
      }, {
         // if some failed we need to send a REVERT event to the children that failed
         cond: ctx => ctx.transitionResponses.every(resp => resp.status !== 'pending')
                        && ctx.transitionResponses.some(resp => resp.status === 'error'),
         actions: pure((ctx) => ctx.transitionResponses.filter(resp => resp.status === 'error').map(resp => send('REVERT', { to: resp.id }))
         target: 'idle'
      }]
   },
}

// and the child machine
{
 initial: 'X'
 states: {
   X: { on: { GO_TO_Y: { target: 'Y', actions: sendParent('RESPONSE', { status: 'success' }) } } },
   Y: { on: { REVERT: 'X' } },
   Z: { on: { GO_TO_Y: { actions: sendParent('RESPONSE', { status: 'error' }) } }}
 }
}
```

Maybe my actor version is completely off, but that's the best I could come up with. It feels very bad compared with the "dirty" method of using the child reference we have to query the mutable state.

Am I completely off-base with my examples above?

PS: I did not do something like "parent asks children for their status and then executes the transition," because that is unsafe - it's possible that a child's state could have changed between the parent asking, verifying, and issuing it's command to transition. Because of this, in the actor model we're forced to do a command "attempt to go to Y and give me a status code", and then a separate REVERT if any attempts failed. 

Overall, it feels like if we're forced to always use a pure actor model we'll very quickly run into a lot of the worst parts of distributed systems: Consensus between nodes is VERY HARD to maintain. Doing pure actor model when we already have a reference to the underlying JS object /seems/ to introduce unnecessary complexity. Unless I'm missing something very obvious that would allow us to simplify the above example?

**Reply by davidkpiano** [reply]
> Overall, it feels like if we're forced to always use a pure actor model we'll very quickly run into a lot of the worst parts of distributed systems: Consensus between nodes is VERY HARD to maintain.

Right, which is why `{ sync: true }` exists, and you _should_ use it. IMO it's safe to read "mutable state" if you are _notified_ when that state changes.

We're still figuring out the API but the only change is going to be something like this:

```diff
SOME_EVENT: {
  target: 'somewhere',
- cond: ctx => ctx.subMachineRef.state.matches('whatever')
+ guard: ctx => ctx.subMachineRef.getSnapshot().matches('whatever')
}
```

That way, you can read the "snapshot" (cached latest value emitted from actor) without dealing with implementation details, and it'll be the same for promise actors, observable actors, etc. (these should all look the same).

**Reply by VanTanev** [reply]
I very much like the `getSnapshot()` method name, it communicates very clearly that we're dealing with a slice-in-time data.

**Comment by tomByrer** [comment]
> can be in either `readonly` or `edit` mode

While testing out how to 'disable' form components, I came upon this trick, where you set up a parallel `disabler` state, then add to the transition statement `in: '#machineID.disabler.enabled',`

https://xstate.js.org/viz/?gist=ca7ca265e8a8c7498a43f7888a14f10f

**Reply by VanTanev** [reply]
That's definitely a great trick, thanks for sharing!
